### ISSUE TITLE
Make /api/v3/addons/compat-override/ less direct DB dependent

### ISSUE THREAD

**Issue by bqbn** [mozilla/addons-server#8768]
### Describe the problem and steps to reproduce it:
It is mentioned in #amo channel that the `/api/v3/addons/compat-override/` is not using ES but hitting the DB slaves directly for its results.

As of today, the `/api/v3/addons/compat-override/` accounts for about 38.6% of the total SAMO requests, which translates into roughly 1.5K requests/second during peak hours.

### What happened?
The cache-machine has been removed in today's release. After the push, we have seen noticeable CPU usage increase on the 3 DB slaves (even during non peak hours, peak hour data yet to come).

Because the request rate of SAMO is typically 10 to 15 times higher than that of AMO, we can probably expect the RDS CPU usage to drop a lot if we can make `/api/v3/addons/compat-override/` to use ES, like other /api/v3/ endpoints do.

### What did you expect to happen?
This is to request we look into the possibility of making `/api/v3/addons/compat-override/` use ES to return results instead of querying DB slaves directly.

**Comment by diox** [comment]
Instead of ES it should just be cached through memcached IMHO. We don't have to do anything super fancy about invalidation, just pick a reasonable time for caching - I suspect that most add-ons have no override and never will be, and for those which have some, it does not change very often.

**Comment by eviljeff** [comment]
yes, compat overrides are not common, and even less common with webextensions. They would have to be stored per guid though - it's not like the lang tools endpoint where it's a list query.  ES seems like the "right" place for it imo, but I agree it's more work to implement.

**Comment by muffinresearch** [comment]
Could this be handled with http caching upstream of the app-server if the content isn't going to change that often?

**Comment by EnTeQuAk** [comment]
We already set proper HTTP cache headers

**Comment by muffinresearch** [comment]
@EnTeQuAk what are they set to?

**Comment by EnTeQuAk** [comment]
> @EnTeQuAk what are they set to?

Sorry, I kinda lied. We are simply re-using ETags as we do all over the place which doesn't necessarily make this less heavy on the database.

Also, simply caching this via HTTP isn't too easy for the same reason this isn't easy to cache with memcached: Firefox can send multiple GUIDs and those in various orderings and the potential complexity comes from the different GUID combinations various Firefoxes may send. So we end up caching data duplicated and inefficient.

What we just agreed on in IRC: Let's keep the PR simple, there are two huge optimizations on the database level already (remove transaction management on that endpoint and avoid count queries for pagination). The endpoint already is crazy fast so we simply have to reduce database-load a bit here which those two will most certainly achieve.

If that doesn't cut it we'll add caching.

**Comment by bqbn** [comment]
We released https://github.com/mozilla/addons-server/pull/8809 last Thursday. It looks like it didn't affect the CPU utilization on the read replicas very much.

Average CPU utilization for the replicas has been consistently in the range of 4-9% (depending on time of the day) since the 7/5 cache-machine removal. This represents 200-300% increase of CPU utilization comparing to before cache-machine removal. However, from an ops perspective, I'm not super worried about it at this point, because a) the performance of this metric has been consistent and doesn't behave like it's unpredictable, and b) we have the option to add one more replica when needed.

Do you see any slowness in any app metrics? If those perform well, I don't think this is as urgent as we originally thought. That said, adding caching for the read replicas as appropriate is always welcomed. :)

**Comment by eviljeff** [comment]
We could just cache the list of all the guids that have overrides to reduce the db load, so only matching ones involve hitting the db.

**Comment by diox** [comment]
Re-opening because of @eviljeff 's caching suggestion and also because I noticed a simple optimization we could make.

`CompatOverride`'s `transformer()` looks like this:
```python
    @staticmethod
    def transformer(overrides):
        if not overrides:
            return

        id_map = {override.id: override for override in overrides}
        qs = CompatOverrideRange.objects.filter(compat__in=id_map)

        for compat_id, ranges in sorted_groupby(qs, 'compat_id'):
            id_map[compat_id].compat_ranges = list(ranges)
```

Notice it does not preset `compat_ranges` if there wasn't any for that override. I'm not sure it happens in the real life, but it's worth investigating. It's quite simple to optimize that if it's happening, [we recently did it for the previews transformer](https://github.com/mozilla/addons-server/blob/de2dc29f7a22ea054ed90bf0e3df422754d2312f/src/olympia/addons/models.py#L1051-L1054).
